reproducibility/codalab/util/analysis_utils.py

```diff
@@ -0,0 +1,120 @@
+                        os.path.join(log_url, f"{id_log_file}.csv")
+                    )
+                except:
+                    print(f"{id_log_file} not available at {log_url}")
+
+            result_dfs[f"{split}_{log_type}"] = pd.read_csv(log_csv_path)
+    return result_dfs
+
+
+def compile_results(dataset, results, in_distribution_val=False):
+    metrics = get_metrics(dataset)
+    compiled_results = dict()
+
+    for algorithm, result_dfs in results.items():
+        compiled_results[algorithm] = dict()
+        print(f"\ndataset={dataset}, algorithm={algorithm}:")
+
+        for result_type in RESULT_TYPES:
+            if len(result_dfs) > 0 and result_type in result_dfs[0]:
+                compiled_results[algorithm][result_type] = dict()
+                for metric in metrics:
+                    compiled_results[algorithm][result_type][metric] = []
+
+        for result_df in result_dfs:
+            for result_type in RESULT_TYPES:
+                if result_type not in result_df:
+                    continue
+
+                for metric in metrics:
+                    sort_metrics = [metric]
+                    # Use the "acc_avg" metric as a tie-breaker for datasets that have "acc_avg"
+                    if metric != "acc_avg" and dataset in [
+                        "amazon",
+                        "camelyon17",
+                        "fmow",
+                        "iwildcam",
+                    ]:
+                        sort_metrics.append("acc_avg")
+
+                    val_split = "id_val_eval" if in_distribution_val else "val_eval"
+                    result = get_early_stopped_row(
+                        result_df[result_type],
+                        result_df[val_split],
+                        sort_fields=sort_metrics,
+                    )
+                    compiled_results[algorithm][result_type][metric].append(
+                        result[metric]
+                    )
+
+        for result_type in RESULT_TYPES:
+            if result_type in compiled_results[algorithm]:
+                for metric in metrics:
+                    # Take the standard deviation and mean across replicants for each metric
+                    compiled_results[algorithm][result_type][f"{metric}_std"] = np.std(
+                        compiled_results[algorithm][result_type][metric]
+                    )
+                    compiled_results[algorithm][result_type][metric] = np.mean(
+                        compiled_results[algorithm][result_type][metric]
+                    )
+    return compiled_results
+
+
+def get_early_stopped_row(src_df, val_df, sort_fields, ascending=False):
+    epoch = val_df.sort_values(sort_fields, ascending=ascending).iloc[0]["epoch"]
+    print(f"Early stopped at {epoch}.")
+    selected_row = src_df.loc[src_df["epoch"] == epoch]
+    assert selected_row.shape[0] == 1
+    return selected_row.iloc[0]
+
+
+def get_metrics(dataset):
```

Comment:
For datasets that have multiple metrics, how are we handling early stopping?

When we report, we should choose one metric to early stop on, and then report the other metrics for that single model. (As opposed to early stopping individually on each metric.)

Reply:
I placed the metric we used for hyperparameter tuning as the first element of the list and then passed it into Panda's sort_values. I added a comment to make this clear.